### PR TITLE
rbd: export/import image-meta when we export/import an image

### PR DESCRIPTION
**Pull request by PCzhangPC**
when we do exporting/importing an image, we should export or import the image-meta of this image at the same time

Signed-off-by: PCzhangPC <pengcheng.zhang@easystack.cn>

### CODE REVIEW

**Comment by PCzhangPC** [comment]
@yangdongsheng please help to review this, thx

File: qa/workunits/rbd/import_export.sh [hunk 1]
```diff
@@ -79,10 +79,18 @@ if rbd help export | grep -q export-format; then
     dd if=/bin/dd of=${TMPDIR}/img bs=1k count=10 seek=100
     rbd import $RBD_CREATE_ARGS ${TMPDIR}/img testimg
     rbd snap create testimg@snap
+    rbd image-meta set testimg key1 value1
+    KEY=`rbd image-meta list testimg|grep key1`
+    VALUE=`rbd image-meta list testimg|grep value1`
```
Comment (yangdongsheng) [inline qa/workunits/rbd/import_export.sh]: The $KEY and $VALUE seems refer to the same line. right? because the key1 and value1 should be outputted in the same line.

File: src/tools/rbd/action/Import.cc [hunk 2]
```diff
@@ -555,6 +555,71 @@ static int decode_and_set_image_option(int fd, uint64_t imageopt, librbd::ImageO
...
+  string key;
+  string value;
+
+  while (r == 0) {
+    __u8 tag;
+    uint64_t length;
+    r = read_tag(fd, RBD_EXPORT_IMAGEMETA_END, import_format, &tag, &length);
+
+    if (r < 0 || tag == RBD_EXPORT_IMAGEMETA_END) {
+      break;
```
Comment (yangdongsheng) [inline src/tools/rbd/action/Import.cc]: if r < 0, this break will lead to ```return 0``` in line 620.

File: qa/workunits/rbd/import_export.sh [hunk 3]
```diff
@@ -79,10 +79,16 @@ if rbd help export | grep -q export-format; then
     dd if=/bin/dd of=${TMPDIR}/img bs=1k count=10 seek=100
     rbd import $RBD_CREATE_ARGS ${TMPDIR}/img testimg
     rbd snap create testimg@snap
+    rbd image-meta set testimg key1 value1
+    KEY=`rbd image-meta list testimg|grep key1`
     rbd export --export-format 2 testimg ${TMPDIR}/img_v2
     rbd import --export-format 2 ${TMPDIR}/img_v2 testimg_import
     rbd info testimg_import
     rbd info testimg_import@snap
+    IMP_KEY=`rbd image-meta list testimg_import|grep key1`
```
Comment (yangdongsheng) [inline qa/workunits/rbd/import_export.sh]: I prefer to check the whole output of ```image-meta list``` rather than check a specified one.

**Comment by PCzhangPC** [comment]
update already,please help to review @yangdongsheng @dillaman

**Comment by yangdongsheng** [comment]
update https://github.com/ceph/ceph/blob/master/doc/dev/rbd-export.rst please

**Comment by PCzhangPC** [comment]
@yangdongsheng update already

File: src/tools/rbd/Utils.h [hunk 4]
```diff
@@ -52,6 +52,10 @@ static const std::string RBD_DIFF_BANNER_V2 ("rbd diff v2\n");
 #define RBD_EXPORT_IMAGE_STRIPE_COUNT	'C'
 #define RBD_EXPORT_IMAGE_END		'E'
 
+#define RBD_EXPORT_IMAGEMETA_KEY		'K'
+#define RBD_EXPORT_IMAGEMETA_VALUE		'V'
+#define RBD_EXPORT_IMAGEMETA_END		'D'
+
```
Comment (yangdongsheng) [inline src/tools/rbd/Utils.h]: I prefer to put the image-meta related into header. That means we have only two sections in the exported file, HEADER and DIFFs. So we can put the metadatas into HEADER section.

Comment (PCzhangPC) [inline src/tools/rbd/Utils.h]: sounds good. and update already

File: src/tools/rbd/action/Import.cc [hunk 5]
```diff
@@ -555,7 +555,45 @@ static int decode_and_set_image_option(int fd, uint64_t imageopt, librbd::ImageO
   return 0;
 }
 
-static int do_import_header(int fd, int import_format, uint64_t &size, librbd::ImageOptions& opts)
+static int read_imagemeta_key_or_value(int fd, string &val, uint64_t length)
```
Comment (yangdongsheng) [inline src/tools/rbd/action/Import.cc]: can we reuse util::read_string() for it?

File: src/tools/rbd/action/Import.cc [hunk 6]
```diff
@@ -555,7 +555,45 @@ static int decode_and_set_image_option(int fd, uint64_t imageopt, librbd::ImageO
...
+                               std::map<std::string, std::string> &pairs)
+{
+  int r = 0;
+
+  //v1 format
+  if (import_format == 1) {
+    return r;
+  }
+
+  if (!pairs.empty()) {
```
Comment (vshankar) [inline src/tools/rbd/action/Import.cc]: same nit as above

File: src/tools/rbd/action/Export.cc [hunk 7]
```diff
@@ -424,6 +424,49 @@ static int do_export_v2(librbd::Image& image, librbd::image_info_t &info, int fd
...
+  std::map<std::string, bufferlist> pairs;
+
+  r = image.metadata_list("", 0, &pairs);
+  if (r < 0) {
+    std::cerr << "failed to retrieve metadata of image : " << cpp_strerror(r)
+              << std::endl;
+    return r;
+  }
+
+  if (!pairs.empty()) {
```
Comment (PCzhangPC) [inline src/tools/rbd/action/Export.cc]: ok,will update soon,thx

File: src/tools/rbd/action/Import.cc (hunk 5 above)
Comment (PCzhangPC) [inline src/tools/rbd/action/Import.cc]: update already, please review this again

File: src/tools/rbd/action/Import.cc [hunk 8]
```diff
@@ -594,6 +618,11 @@ static int do_import_header(int fd, int import_format, uint64_t &size, librbd::I
       r = decode_and_set_image_option(fd, RBD_IMAGE_OPTION_STRIPE_UNIT, opts);
     } else if (tag == RBD_EXPORT_IMAGE_STRIPE_COUNT) {
       r = decode_and_set_image_option(fd, RBD_IMAGE_OPTION_STRIPE_COUNT, opts);
+    } else if (tag == RBD_EXPORT_IMAGE_META_KEY) {
+      r = utils::read_string(fd, 4096, &key);
+    } else if (tag == RBD_EXPORT_IMAGE_META_VALUE) {
+      r = utils::read_string(fd, 4096, &value);
```
Comment (yangdongsheng) [inline src/tools/rbd/action/Import.cc]: 4096? or length?

**Comment by PCzhangPC** [comment]
@ceph-jenkins retest this please

**Comment by yangdongsheng** [comment]
Hi @dillaman  This looks good to me. what's your opinion?

**Comment by PCzhangPC** [comment]
@dillaman please help to review this pr too, thx

File: doc/dev/rbd-export.rst [hunk 9]
```diff
@@ -65,6 +65,20 @@ Image Stripe count
 - le64: length of appending data (8)
 - le64: image striping count
 
+ImageMeta Key
```
Comment (dillaman) [inline doc/dev/rbd-export.rst]: Image-meta should be combined into a single tag -- so perhaps the 'M' tag (for image-**m**eta), le64 for tag length as usual, then the string key followed by the string value.

File: src/tools/rbd/action/Export.cc [hunk 10]
```diff
@@ -414,6 +414,36 @@ static int do_export_v2(librbd::Image& image, librbd::image_info_t &info, int fd
   ::encode(length, bl);
   ::encode(stripe_count, bl);
 
+  //encode imageMeta key and value
+  std::map<std::string, bufferlist> imagemetas;
+
+  r = image.metadata_list("", 0, &imagemetas);
```
Comment (dillaman) [inline src/tools/rbd/action/Export.cc]: This needs to be a loop -- read X key/value pairs at a time and continue the loop if the number of retrieved pairs equals the request count.

File: src/tools/rbd/action/Import.cc [hunk 11]
```diff
@@ -555,7 +555,28 @@ static int decode_and_set_image_option(int fd, uint64_t imageopt, librbd::ImageO
   return 0;
 }
 
-static int do_import_header(int fd, int import_format, uint64_t &size, librbd::ImageOptions& opts)
+static int do_import_imagemeta(int import_format, librbd::Image& image,
```
Comment (dillaman) [inline src/tools/rbd/action/Import.cc]: Nit: ```imagemeta``` -> ```metadata```

File: src/tools/rbd/action/Export.cc (hunk 10 above)
Comment (PCzhangPC) [reply src/tools/rbd/action/Export.cc]: no need to be a loop in fact. `image.metadata_list("", 0, &imagemetas)` will fetch all key-value pairs into the map `imagemetas`  if we set the second parameter as 0.   Or am i missing sth ?

Comment (dillaman) [reply src/tools/rbd/action/Export.cc]: It works for a large number of keys but not all possible (worse-case): https://github.com/ceph/ceph/blob/master/src/cls/rbd/cls_rbd.cc#L2654

Comment (yangdongsheng) [reply src/tools/rbd/action/Export.cc]: Hi @dillaman , if we specify the max_return to 0. I think the code you pointed is trying to read all metadatas in a loop until there is no ```more```.

Comment (dillaman) [reply src/tools/rbd/action/Export.cc]: It returns a maximum of 64 keys if zero is provided

Comment (PCzhangPC) [reply src/tools/rbd/action/Export.cc]: if we set `max_return` as 0,  the `max_read` will be set to `RBD_MAX_KEYS_READ(64)`. but it only affect the the function `cls_cxx_map_get_vals` , the loop will break if no more image-meta ,which means that we get all metadatas.  so, when will the worse-case happen?

Comment (dillaman) [reply src/tools/rbd/action/Export.cc]: If I tell cls_cxx_map_get_vals to retrieve a maximum of 5 entries and it returns 5 and says more are available, that's a bug. The more bool is supposed to be used to indicate a partial set being retrieved.

Comment (yangdongsheng) [reply src/tools/rbd/action/Export.cc]: @dillaman  I have a try to fix the bug you mentioned.
```
@@ -6418,13 +6419,15 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
          tracepoint(osd, do_osd_op_pre_omapgetvals, soid.oid.name.c_str(), soid.snap.val, "???", 0, "???");
          goto fail;
        }
+
+       bool truncated = false;
        if (max_return > cct->_conf->osd_max_omap_entries_per_request) {
          max_return = cct->_conf->osd_max_omap_entries_per_request;
+         truncated = true;
        }
        tracepoint(osd, do_osd_op_pre_omapgetvals, soid.oid.name.c_str(), soid.snap.val, start_after.c_str(), max_return, filter_prefix.c_str());
 
        uint32_t num = 0;
-       bool truncated = false;
        bufferlist bl;
        if (oi.is_omap()) {
          ObjectMap::ObjectMapIterator iter = osd->store->get_omap_iterator(
@@ -6437,12 +6440,12 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
          iter->upper_bound(start_after);
          if (filter_prefix > start_after) iter->lower_bound(filter_prefix);
          for (num = 0;
-              iter->valid() &&
+              num <= max_return &&
+                iter->valid() &&
                 iter->key().substr(0, filter_prefix.size()) == filter_prefix;
               ++num, iter->next(false)) {
            dout(20) << "Found key " << iter->key() << dendl;
-           if (num >= max_return ||
-               bl.length() >= cct->_conf->osd_max_omap_bytes_per_request) {
+           if (bl.length() >= cct->_conf->osd_max_omap_bytes_per_request) {
              truncated = true;
              break;
            }
```

But I have two questions about it:
(1) I am a little concern about changing this behavior of it if there is some other guys rely on this ```bug```. e.g ```rbd image-meta list```. maybe more. it would make a regression. 

(2) Actually, we can check the ```max_return``` and ```vals.size()``` to know did we only retrieved a partial set. And we can try again with a specified last_read. Then we don't need another parameter of ```more``` to tell us that. Just my opinion. 
Is there a design document for ```CEPH_OSD_OP_OMAPGETVALS``` op? 

Thanx

Comment (dillaman) [reply src/tools/rbd/action/Export.cc]: (1) ```rbd image-meta list``` is broken and should be fixed [1]
(2) No such documentation -- the code is the documentation. The purpose is for case where the OSD cannot immediately return all entries but (1) more exist and (2) the ```max_return``` hasn't been met yet so the caller should try again (like receiving an -EAGAIN on a syscall).

[1] http://tracker.ceph.com/issues/21179

Comment (yangdongsheng) [reply src/tools/rbd/action/Export.cc]: Thanx @dillaman , will open a PR for this.

File: doc/dev/rbd-export.rst (hunk 9 above)
Comment (PCzhangPC) [reply doc/dev/rbd-export.rst]: what you suggest is that we should combined all key-value paires into one tag?

File: src/tools/rbd/action/Export.cc (hunk 10 above)
Comment (PCzhangPC) [reply src/tools/rbd/action/Export.cc]: @dillaman so, what should be done are as follow:  
1. fix the bug you mention, make `more` indicate a partial set being retrieved.
2. do some changes at where the `cls_cxx_map_get_vals` function is being used in `cls_rbd.cc` :  realize a new way to fetch all date instead of using `more` as a loop flag now

Comment (dillaman) [reply src/tools/rbd/action/Export.cc]: I'm not sure I am following your last commit:
(1) the tracker ticket is to just update the rbd CLI to iterate over partial sets -- ```more``` already indicates that the OSD had to truncate the result but the requested limit has not been reached. See the ```osd_max_omap_bytes_per_request``` configuration value that can cause the truncation of the results.
(2) The ```more``` boolean should already be handled everywhere.

Comment (PCzhangPC) [reply src/tools/rbd/action/Export.cc]: hi @dillaman ,there are sth i am confused with. 
1. `image.metadata_list()`  is not broken currently. I have tried 300 metadates and succeed to get all.  (i think it is the 'bug' that makes it work)
2.  if we fix the bug you mentioned , which is that `more` should be `false` if we tell `cls_cxx_map_get_vals` to retrieve 5 entries and it returns 5 (am i right ?). But, please consider about this:  we have 65 metadates and want to get all.  The first time `cls_cxx_map_get_vals` succeeds to fetch 64 pairs(`RBD_MAX_KEYS_READ`) and `more` is set as `false`,  so it will not go to fetch the last pair, which means image.imagemeta_list() would be broken after fix. And that's why i said "realize a new logic to fetch all metadates instead of using more as a loop flag" in my last comment.
3. about "iterate over `image.metadata_list`": no matter how many time `image.metadata_list()` we call, the result of each time will be repeated.  `image.metadata_list()` will start from the ‘first‘ pair but not from the end of last function call to get metadates.

Comment (dillaman) [reply src/tools/rbd/action/Export.cc]: OK -- looks like that was recently broken and is a totally separate issue. I'll push a fix for the broken handling of ```more``` [1].

[1] http://tracker.ceph.com/issues/21247

**Comment by PCzhangPC** [comment]
@ceph-jenkins retest this please

**Comment by PCzhangPC** [comment]
@dillaman updated

File: qa/workunits/rbd/import_export.sh [hunk 12]
```diff
@@ -79,10 +79,16 @@ if rbd help export | grep -q export-format; then
...
     rbd import $RBD_CREATE_ARGS ${TMPDIR}/img testimg
     rbd snap create testimg@snap
+    rbd image-meta set testimg key1 value1
+    IMAGEMETA_BEFORE=`rbd image-meta list testimg`
     rbd export --export-format 2 testimg ${TMPDIR}/img_v2
     rbd import --export-format 2 ${TMPDIR}/img_v2 testimg_import
     rbd info testimg_import
     rbd info testimg_import@snap
+    IMAGEMETA_AFTER=`rbd image-meta list testimg_import`
+    if [ "$IMAGEMETA_BEFORE" != "$IMAGEMETA_AFTER" ]; then
```
Comment (dillaman) [inline qa/workunits/rbd/import_export.sh]: Nit: ```[ "$IMAGEMETA_BEFORE" = "$IMAGEMETA_AFTER" ]```

File: src/tools/rbd/action/Import.cc [hunk 13]
```diff
@@ -594,6 +618,16 @@ static int do_import_header(int fd, int import_format, uint64_t &size, librbd::I
       r = decode_and_set_image_option(fd, RBD_IMAGE_OPTION_STRIPE_UNIT, opts);
     } else if (tag == RBD_EXPORT_IMAGE_STRIPE_COUNT) {
       r = decode_and_set_image_option(fd, RBD_IMAGE_OPTION_STRIPE_COUNT, opts);
+    } else if (tag == RBD_EXPORT_IMAGE_META) {
+      r = utils::read_string(fd, length, &key);
```
Comment (dillaman) [inline src/tools/rbd/action/Import.cc]: Nit: why not follow the existing pattern of just calling a helper function to read the associated values of the tag struct and update the image (avoiding the intermediate list)?

Comment (PCzhangPC) [reply src/tools/rbd/action/Import.cc]: we havn't create an import_image yet. so what we could do is just to collect them all and set image-meta later.

**Comment by PCzhangPC** [comment]
@ceph-jenkins retest this pls

**Comment by PCzhangPC** [comment]
@dillaman updated

File: src/tools/rbd/action/Import.cc [hunk 14]
```diff
@@ -555,7 +555,47 @@ static int decode_and_set_image_option(int fd, uint64_t imageopt, librbd::ImageO
...
+       it != imagemetas.end(); ++it) {
+    r = image.metadata_set(it->first, it->second);
+    if (r < 0)
+      return r;
+  }
+
+  return r;
+}
+
+static int decode_imagemeta(int fd, uint64_t length, std::map<std::string,
```
Comment (dillaman) [inline src/tools/rbd/action/Import.cc]: Nit: don't break type between multiple lines and out params should be passed by pointer

File: src/tools/rbd/action/Import.cc [hunk 15]
```diff
@@ -555,7 +555,47 @@ static int decode_and_set_image_option(int fd, uint64_t imageopt, librbd::ImageO
...
+  r = utils::read_string(fd, length, &key);
+  if (r < 0)
+    return r;
+
+  r = utils::read_string(fd, length, &value);
+  if (r < 0)
+    return r;
+
+  imagemetas[key] = value;
+  return r;
```
Comment (dillaman) [inline src/tools/rbd/action/Import.cc]: Nit: for clarity, ```return 0```

File: src/tools/rbd/action/Import.cc (hunk 14 above)
Comment (PCzhangPC) [reply src/tools/rbd/action/Import.cc]: my bad. updated already

File: src/tools/rbd/action/Import.cc [hunk 16]
```diff
@@ -555,7 +555,46 @@ static int decode_and_set_image_option(int fd, uint64_t imageopt, librbd::ImageO
...
+       it != imagemetas.end(); ++it) {
+    r = image.metadata_set(it->first, it->second);
+    if (r < 0)
+      return r;
+  }
+
+  return r;
+}
+
+static int decode_imagemeta(int fd, uint64_t length, std::map<std::string, std::string> &imagemetas)
```
Comment (dillaman) [inline src/tools/rbd/action/Import.cc]: Nit: ```imagemetas``` should be passed by pointer

File: src/tools/rbd/action/Import.cc [hunk 17]
```diff
@@ -555,7 +555,46 @@ static int decode_and_set_image_option(int fd, uint64_t imageopt, librbd::ImageO
   return 0;
 }
 
-static int do_import_header(int fd, int import_format, uint64_t &size, librbd::ImageOptions& opts)
+static int do_import_metadata(int import_format, librbd::Image& image,
+                              std::map<std::string, std::string> &imagemetas)
```
Comment (dillaman) [inline src/tools/rbd/action/Import.cc]: Nit: ```imagemetas``` should be passed by const reference

File: src/tools/rbd/action/Import.cc [hunk 18]
```diff
@@ -578,7 +617,7 @@ static int do_import_header(int fd, int import_format, uint64_t &size, librbd::I
     opts.set(RBD_IMAGE_OPTION_FORMAT, image_format);
   }
 
-  while (r == 0) {
+  while (r >= 0) {
```
Comment (dillaman) [inline src/tools/rbd/action/Import.cc]: Nit: was this change required?

File: src/tools/rbd/action/Import.cc (hunk 16 above)
Comment (PCzhangPC) [reply src/tools/rbd/action/Import.cc]: i'am not very sure. any bad effect if passed by reference ?

Comment (vshankar) [reply src/tools/rbd/action/Import.cc]: @PCzhangPC its just a convention -- if an argument is modified by a routine then pass it as a pointer.

Comment (PCzhangPC) [reply src/tools/rbd/action/Import.cc]: ok. thank you

**Comment by PCzhangPC** [comment]
@dillaman  need review,thx

File: src/tools/rbd/action/Import.cc [hunk 19]
```diff
@@ -555,7 +555,46 @@ static int decode_and_set_image_option(int fd, uint64_t imageopt, librbd::ImageO
...
+  r = utils::read_string(fd, length, &value);
+  if (r < 0)
+    return r;
+
+  (*imagemetas)[key] = value;
+  return 0;
+}
+
+static int do_import_header(int fd, int import_format, uint64_t &size, librbd::ImageOptions& opts,
+                            std::map<std::string, std::string> &imagemetas)
```
Comment (dillaman) [inline src/tools/rbd/action/Import.cc]: Nit: pass out-param as pointer instead of by reference

**Comment by PCzhangPC** [comment]
@dillaman need review，THX
